### ISSUE TITLE
Unable to set HTTP_MAX_HEADER_LENGTH

### ISSUE THREAD

**Issue by skyeagle** [boazsegev/iodine#99]
### System Information

 - **OS**:             macOS Catalina 10.15.6 
 - **Ruby**:         2.6.5
 - **Version**:     0.7.42

### Description

Unable to set HTTP_MAX_HEADER_LENGTH neither from CLI nor through compile options. 

### Testing code

starting the server
```sh
iodine -p 3000 -w 1 -maxhd 32768
```
in another console
```
curl -H "Cookie: _session=$(printf '=%.0s' $(seq 1 10000));" http://localhost:3000
# => Payload Too Large
```

same result with compile options:
```
export CFLAGS="-DHTTP_MAX_HEADER_LENGTH=32768"
gem pristine iodine
```
### Expected behavior

It should allow setting the request's maximum header length. 

### Actual behavior

The server doesn't seem to allow to change that setting and responds with `HTTP 413 Payload Too Large` anyways.

Thank you in advance for the fantastic work on this server and for helping with this issue!

**Comment by skyeagle** [comment]
Just to clarify, it doesn't work with passing it in kilobytes as well.

**Comment by boazsegev** [comment]
Hi @skyeagle and thank you for opening this issue.

I need some more information so I could track down the issue. I hope to resolve it for you, but you should know that most **intermediaries** (proxies, active load balancers, etc') **will limit each header's length**. The most common setting in the wild is to **limit each header length to 4Kb**. Some intermediaries will allow 8Kb to pass through.

Just so we're clear... the `-maxhd` value is supposed to be the limit per request - that's the length of all HTTP headers combined (names + values), not the maximum length of each header separately.

In addition, each single header (name + value) must be less than the hardcoded `HTTP_MAX_HEADER_LENGTH` (as set during compile time) and all headers together should be less than `-maxhd`.

**Do these settings fail on your local machine** (no intermediaries / middleware)?

If these don't work locally , there might be an issue with the code and I will need to look deeper.

If the issue occurs behind a proxy or with validation middleware, please make sure that you don't have a limit set in the proxy and that your middleware isn't verifying that the header length is suitable for traversing the web (4Kb would be the common limit imposed by such middleware).

Looking forward for further details.

Thanks again,
   Bo.

**Comment by boazsegev** [comment]
Update:

Seems to be a `gem install` issue / usage concern...

Testing on my machine it appears that the `-maxhd` value works correctly when making the maximum header value smaller or larger. It does not effect the per-header limit, only the total-header limit.

Compiling with a custom `HTTP_MAX_HEADER_LENGTH` succeeds and allows the `curl` example you provided to work.

However, **using `CFLAGS="..."` is totally useless** and it appears to be ignored by the Ruby makefile for some reason.

[The correct way to install iodine with updated CFLAGS](https://stackoverflow.com/questions/6761670/is-there-any-way-to-change-gcc-compilation-options-for-a-gem) appears to be:

```bash
gem install iodine-- --with-cflags=\"-DHTTP_MAX_HEADER_LENGTH=48000\"
```

At which point you can run iodine using (value is in Kb, not bytes):

```bash
iodine -p 3000 -w 1 -maxhd 32
```

Works on my machine (macOS).

**Comment by skyeagle** [comment]
hey @bjeanes, thanks for the blazing-fast response and clarification.

We hit this issue on our infrastructure first on Linux machines behind an AWS load balancer where we are trying to replace puma+nginx with iodine only. So, it's reproducible locally (multiple macosx) and on Linux boxes.

It seems like this issue has to do with the size/count of cookies we have. One of the AWS ELB auth cookies is around 4k.

In the case of Nginx, we needed to adjust [large_client_header_buffers](http://nginx.org/en/docs/http/ngx_http_core_module.html#large_client_header_buffers) to 32k to make it work. For iodine, as a temporary workaround, I've made [a fork](https://github.com/pioneerworks/iodine/commit/15ee6ea99845095b30c04cd3830e6fb216fc4860) that seemed to work locally.

While I was wrapping up my response you already replied and I tried things you suggested...

this works
```sh
gem install iodine-- --with-cflags=\"-DHTTP_MAX_HEADER_LENGTH=48000\"
```

this doesn't work
```
iodine -p 3000 -w 1 -maxhd 32
```

Again, thanks for getting back and such valuable feedback!
